### ISSUE TITLE
Unable to use kfp.Client() from local / VM

### ISSUE THREAD

**Issue by devashishshankar** [kubeflow/pipelines#4143]
### What steps did you take:

I am simply trying to list pipelines / upload pipelines using kfp.Cilent() locally. I've tried different authentication options. Sometimes, I get an auth error - other times I get an empty response. I will summarize what all I've tried here.

A note on environment: I had done a gcloud auth login locally (More details on env in Environment section) 

#### 1. Pass only host

```
client = kfp.Client(host='https://<my-kf>.endpoints.<my-proj>.cloud.goog')
print(client.list_pipelines())

```

I had expected this to work because of the gcloud auth login.

Response:
```
Reason: Unauthorized
HTTP response headers: HTTPHeaderDict({'X-Goog-IAP-Generated-Response': 'true', 'Content-Length': '36', 'Content-Type': 'text/html; charset=UTF-8', 'Date': 'Fri, 03 Jul 2020 14:37:09 GMT', 'Alt-Svc': 'clear'})
HTTP response body: Invalid IAP credentials: empty token
```

#### 2. Pass client id

```
client = kfp.Client(host='https://<my-kf>.endpoints.<my-proj>.cloud.goog',  client_id=<my_iap_client_id>)
print(client.list_pipelines())

```

I tried this as I saw some code that tries to auto get credentials from sa when client_id is passed. I don't receive an 
Authorization error - but an empty response.

```
{'next_page_token': None, 'pipelines': None, 'total_size': None}
```

This is incorrect because I have a few pipelines defined.

#### 3. Pass other_client_id and other_client_secret

This was done by creating a new IAP credential. I tried it because it solved the issue for people in https://github.com/kubeflow/pipelines/issues/2954 

On running this, it redirects me to a web URL to authenticate. Once done - I still get the same empty response as 2.

```
{'next_page_token': None, 'pipelines': None, 'total_size': None}
```

#### 4. Pass `gcloud auth print-access-token` as existing_token

(Again the authentication is using my GCP account - with which I can access kubeflow UI)
```
client = kfp.Client(host='https://<my-kf>.endpoints.<my-proj>.cloud.goog',
                               existing_token="<token obtained using `gcloud auth print-access-token`>")

```

Response:
```
Reason: Unauthorized
HTTP response headers: HTTPHeaderDict({'X-Goog-IAP-Generated-Response': 'true', 'Content-Length': '36', 'Content-Type': 'text/html; charset=UTF-8', 'Date': 'Fri, 03 Jul 2020 14:52:07 GMT', 'Alt-Svc': 'clear'})
HTTP response body: Invalid IAP credentials: empty token
```

#### 5. existing_token with `client_id`
```
client = kfp.Client(host='https://<my-kf>.endpoints.<my-proj>.cloud.goog',
                               existing_token="<token obtained using `gcloud auth print-access-token`>",
                               client_id=<my_iap_client_id>
)
```

Response:
```
Reason: Unauthorized
HTTP response headers: HTTPHeaderDict({'X-Goog-IAP-Generated-Response': 'true', 'Content-Length': '94', 'Content-Type': 'text/html; charset=UTF-8', 'Date': 'Fri, 03 Jul 2020 14:53:47 GMT', 'Alt-Svc': 'clear'})
HTTP response body: Invalid IAP credentials: Expected JWT to have 3 parts separated by a '.' but there are 2 parts
```

At this point, I figured it needed a JWT token not an Oauth token. Hence I tried generating a JWT token.

#### 6. existing_token (JWT) with `client_id`

```
with open("<my-sa>", "r") as f:
    info = json.load(f)
    credentials_content = (
            (info.get("type") == "service_account") and info or None
    )

    from google.oauth2 import service_account

    credentials = service_account.IDTokenCredentials.from_service_account_info(
        credentials_content, target_audience="<my_client_id>"
    )

credentials.refresh(Request())

client = kfp.Client(host='https://<my-kf>.endpoints.<my-proj>.cloud.goog',
                    existing_token=credentials.token,
                    client_id=<my_client_id>,

```
I gave permission to the service account before running this - otherwise I still got an IAP auth issue.

Response:  (Empty) - Same as 2 & 3.

```
{'next_page_token': None, 'pipelines': None, 'total_size': None}
```

#### TL;DR

I tried various possible ways to authenticate. I got either an Authorization error or an empty response. I think I'm authenticating correctly in  2, 3 & 6 - since I don't get an error. But then why do I get an empty response? Maybe it's something silly which I'm missing? I get the correct response from within the Kubeflow server's notebook. 

I also tried a few of this steps on a GCP VM - with very similar results.

### What happened:

(Described Above)

### What did you expect to happen:

Running the script on the Kubeflow jupyter server, gives a list of pipelines (I have a few defined)

```
client.list_pipelines()
```

### Environment:

* Kubeflow is installed on GCP/ GKE.
* Python version: Python 3.7.6
* I have done `gcloud auth login` on local. I tried `gcloud auth activate-service-account` as well.
* I have configured Kubeflow to use IAP authentication.
* How did you deploy Kubeflow Pipelines (KFP)? - [Deployed using CLI](https://www.kubeflow.org/docs/gke/deploy/deploy-cli/)
* KFP version: I deployed `kfctl_gcp_iap.v1.0.2.yaml`. But on the left side I see `build version dev_local`. I think this is related to https://github.com/kubeflow/kubeflow/issues/5007?

KFP SDK version: 

```
kfp                           0.5.1
kfp-server-api                0.5.0
```

### Anything else you would like to add:

/kind bug

/area sdk

**Comment by devashishshankar** [comment]
This got solved courtesy doit-intl team. 

So, what works (at least in GKE) is connecting to the cluster using `gcloud container clusters get-credentials <cluster-name>`. Once that is done, simply doing `kfp.Client().list_pipelines()` works (no args).

This may be worth mentioning in the python lib documentation somewhere? I had no idea it was actually riding on kubectl authentication. 

Keeping the issue open - since the IAP auth failure may be a potential bug. But if this is indeed the preferred way of connecting - I can close this issue.

**Comment by Bobgy** [comment]
@devashishshankar for IAP, we don't have public documentation, you were very close on 3.

```
import kfp
client = kfp.Client(host='https://<KF_NAME>.endpoints.<PROJECT>.cloud.goog/pipeline', 
client_id='<AAAAAAAAAAAAAAAAAAAAAA>.apps.googleusercontent.com', 
other_client_id='<BBBBBBBBBBBBBBBBBBB>.apps.googleusercontent.com', 
other_client_secret='<CCCCCCCCCCCCCCCCCCCC>')
print(client.list_pipelines())

client.create_experiment(name='<Your experiment name>', namespace='<Your namespace>')
print(client.list_experiments(namespace='<Your namespace>'))
client.run_pipeline(experiment_id='<Your experiment ID>', job_name='<Your job ID>', pipeline_id='<Your pipeline ID>')
print(client.list_runs(experiment_id='<Your experiment ID>'))
print(client.list_runs(namespace='<Your namespace>'))
```

You also need the client_id passed in, it should be your Kubeflow IAP's oauth client id.

**Comment by Bobgy** [comment]
Documentation is already planned: https://github.com/kubeflow/pipelines/issues/2704

**Comment by Bobgy** [comment]
> So, what works (at least in GKE) is connecting to the cluster using gcloud container clusters get-credentials <cluster-name>. Once that is done, simply doing kfp.Client().list_pipelines() works (no args).
This may be worth mentioning in the python lib documentation somewhere? I had no idea it was actually riding on kubectl authentication.

Good point, I'm not sure why it also works this way.
/cc @rmgogogo 
Is this expected behavior?

**Comment by rmgogogo** [comment]
> > So, what works (at least in GKE) is connecting to the cluster using gcloud container clusters get-credentials . Once that is done, simply doing kfp.Client().list_pipelines() works (no args).
> > This may be worth mentioning in the python lib documentation somewhere? I had no idea it was actually riding on kubectl authentication.
> 
> Good point, I'm not sure why it also works this way.
> /cc @rmgogogo
> Is this expected behavior?

Not on "kubectl authentication", it uses "gcloud" CLI to get the token. We had an email before on it.
It has to be on GKE as it uses inverse proxy which only on GKE. for options for IAP path, defer to Yuan, standalone doesn't support IAP.

https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/_auth.py#L40

**Comment by Bobgy** [comment]
We have public documentation on how to connect now: https://www.kubeflow.org/docs/gke/pipelines/authentication-sdk/#connecting-to-kubeflow-pipelines-in-a-full-kubeflow-deployment.
